### PR TITLE
Allow * in the email-domain list.

### PR DESCRIPTION
**Pull request by morarucostel**
Allowing the `*` in the individual email-domain entries.  
Fix for the #927 issue. 

## Description
The `*` sign is replaced by a regex allowing the any email characters to be present. 
This is done only once, when the domains array is created. 
The validator is checking if the regex is valid OR if its a perfect match (allowing for backward compatibility).

## How Has This Been Tested?

Tested in our k8s environments and with the help of unit tests.

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by NickMeves** [comment]
Ping me when this is ready for another look once the security related changes to the architecture we discussed in the linked issue are coded up 👍

**Comment by morarucostel** [comment]
Hi @NickMeves , just pushed the changes to remove the `*` regex based and introduced the sub-domain based on the `.` logic.

File: CHANGELOG.md
Comment (NickMeves) [inline CHANGELOG.md:49]: It looks like this typo got cleaned up in a PR since yours was made. When you rebase off the latest master can you make sure the link typo stays fixed?

File: validator.go
Comment (NickMeves) [inline validator.go:94]: This might be best as a helper method to contain all the complexity in this check.

In the first case (wildcard subdomain denoted by a `.` prefix), I feel like we still need to make sure there's an `@`. Not sure how I would abuse it missing as a hacker, but it _feels_ like I can 😅

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:213]: Wow - I haven't actually looked at this file closely before, these tests haven't been really touched in 5 years and need some TLC.

If you have the bandwidth, can you align your new test cases to the Test Table type tests we have in the rest of the project? Here's a sample: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_default_test.go#L61

File: validator.go
Comment (NickMeves) [reply validator.go:94]: (In what will be the new helper) Maybe use `strings.SplitN(email, "@", 2)` and check the second element with the suffix check?

Not sure if we want explicit `splitN...2` which assumes only a single `@` or if we should just do an uncapped split and just take the last element in the split slice for the suffix check.

File: validator_test.go
Comment (NickMeves) [reply validator_test.go:213]: You can use either ginkgo style in that sample or traditional `t.Assert`, no preference on that. More just want to see any new test cases structured in either test tables or ginkgo.

File: validator.go
Comment (morarucostel) [reply validator.go]: Created a helper method that contains the logic of validating the email agains the emailDomain. 

I choose to use the uncapped split by `@` and to check the last item in order to support backward compatibility (some teams might have used an user that doesn't contain just one `@`).

File: validator_test.go
Comment (morarucostel) [reply validator_test.go]: Refactored the tests to use the provided example.

**Comment by morarucostel** [comment]
@NickMeves Thanks for the review. Can you please have another look at it?

File: validator.go
Comment (NickMeves) [inline validator.go:112]: NIT:

```suggestion
// isEmailValidWithDomains checks if the authenticated email is validated against the provided domain
```

File: validator.go
Comment (NickMeves) [inline validator.go:113]: We should take in `emailDomains []string` instead of `emailDomain string` here and manage the looping in this function.

Then we don't need this logic above:

```
		for _, domain := range domains {
			valid = valid || isEmailValidWithDomains(email, domain)
		}
```

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:304]: I feel like needed cleanup is missing here. Does the existing `NewValidatorTest` and it's helper methods (particularly `TearDown`) work in these tests for handling the emailsFile stuff?

File: validator.go
Comment (morarucostel) [reply validator.go]: Updated

File: validator.go
Comment (morarucostel) [reply validator.go]: Done.

File: validator_test.go
Comment (morarucostel) [reply validator_test.go:304]: Since the new tests are not using/instantiating the `NewValidatorTest`, I believe they are not goint to get cleaned by the `TearDown`. 
I've introduced the `defer os.RemoveAll(dir)` that should clean-up the temporary dir.

File: CHANGELOG.md
Comment (NickMeves) [inline CHANGELOG.md:50]: We should update this CHANGELOG to reflect the compromise functionality we ended up with that didn't have the security issues.

File: validator_test.go
Comment (NickMeves) [reply validator_test.go:304]: Let's get these in the test cases so we aren't risking polluting state between test runs.

```
f, err := ioutil.TempFile("", fmt.Sprintf("%s_", tc.name))
defer os.Remove(f)
g.Expect(err).ToNot(HaveOccurred())
```

And ideally handling errors in the deferred case use a closure in the defer to wrap handling any cleanup errors. But this isn't required, we don't Gosec scan test code. This would be required in non-tests for `defer` cleanup stuff for error handling.
```
defer func(t *testing.T) {
  if err := f.Close(); err != nil {
    t.Fatal(err)
  }
}(t)
```

File: validator.go
Comment (NickMeves) [inline validator.go:111]: Nice! Looks very clean, thanks!

File: validator_test.go
Comment (morarucostel) [reply validator_test.go]: @NickMeves Thanks for the code and the review. I actually didn't know about this type of cleaning-up and defering.

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:309]: Lol - This comment was just for your benefit in the sample -- not intended to go into the codebase 🤣

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:310]: Why a whole `TempDir` instead of just the single `TempFile` you need for the test case?

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:311]: This is redundant if you do the `defer func...` a couple lines lower

File: validator_test.go
Comment (NickMeves) [inline validator_test.go:321]: This should be `0600` permissions otherwise Gosec security scans will complain.

Comment (NickMeves) [reply validator_test.go:321]: In general, what's the rationale for reinventing the wheel in this test case? The `ValidatorTest` struct in this package has helper methods to handle all the proper details of making temporary test files & cleanup (`WriteEmails` and `TearDown` methods).

If that test helper has any weaknesses, I think the proper route is to fix it appropriately so it can be used throughout the test suite still.

Comment (morarucostel) [reply validator_test.go]: Agree, I guess i misread the previous message. 
I've refactored the newly introduced tests to use the existing clean-up methods.

**Comment by morarucostel** [comment]
@NickMeves Just refactored all the test to use the new model.

**Review by NickMeves — DISMISSED** [review]
Thanks for cleaning up the test cases! Looks good

**Comment by NickMeves** [comment]
Hmm, I thought I could merge in master real quick and get clean actions and merge now. Apparently not... 😢 

@morarucostel can you rebase off master (and ditch my merge commit).

Also, if you want to keep your commit history, can you squash some of them and get them aligned to this general style (https://chris.beams.io/posts/git-commit/)? Otherwise I can just squash & merge when the time comes if you don't care. But I don't want the non-ascii emojis in the commit log.

**Comment by morarucostel** [comment]
@NickMeves I think I just finished the rebase and squash.

**Review by NickMeves — APPROVED** [review]

**Comment by NickMeves** [comment]
It looks like the test suite isn't passing with some of the test changes:

```
[2020/12/22 01:15:11] [validator.go:50] failed opening authenticated-emails-file="/tmp/test_auth_emails_796163895", open /tmp/test_auth_emails_796163895: no such file or directory
FAIL	github.com/oauth2-proxy/oauth2-proxy/v7	1.145s
```

**Review by NickMeves — CHANGES_REQUESTED** [review]
It looks like the test suite isn't passing with some of the test changes:
```
[2020/12/22 01:15:11] [validator.go:50] failed opening authenticated-emails-file="/tmp/test_auth_emails_796163895", open /tmp/test_auth_emails_796163895: no such file or directory
FAIL	github.com/oauth2-proxy/oauth2-proxy/v7	1.145s
```

File: validator_test.go
Comment (JoelSpeed) [inline validator_test.go:376]: This makes me feel like we need to update the documentation of this, I had expected that if I wrote `.example.com` that emails to that domain and its subdomains would have been allowed, this is how our whitelist domain works IIRC 🤔 

Secondly, do we have a test case where we do `.example.com` as an allowed domain and the email itself is `something@fooexample.com`, just to make sure that we are including the `.` in the suffix check and don't break that in the future by trying to be clever about something

Comment (NickMeves) [reply validator_test.go:376]: Good catch! You are right -- in `whitelistDomains` it does have a check for `redirectHostname == strings.TrimLeft(domain, ".")` essentially.

We should add that here to match.

**Comment by morarucostel** [comment]
> It looks like the test suite isn't passing with some of the test changes:
> 
> ```
> [2020/12/22 01:15:11] [validator.go:50] failed opening authenticated-emails-file="/tmp/test_auth_emails_796163895", open /tmp/test_auth_emails_796163895: no such file or directory
> FAIL	github.com/oauth2-proxy/oauth2-proxy/v7	1.145s
> ```
@NickMeves Can you please double check on your side?
I ran successfully all the tests locally and also the the github CI-build was successfully completed. 
```
--- PASS: TestValidatorSubDomains (0.05s)
    --- PASS: TestValidatorSubDomains/EmailNotInCorrect1stSubDomainsNotInEmails (0.00s)
    --- PASS: TestValidatorSubDomains/EmailInFirstDomain (0.00s)
    --- PASS: TestValidatorSubDomains/EmailNotInCorrect2ndSubDomainsNotInEmails (0.01s)
    --- PASS: TestValidatorSubDomains/EmailInSecondDomain (0.00s)
    --- PASS: TestValidatorSubDomains/EmailInFirstEmailList (0.00s)
    --- PASS: TestValidatorSubDomains/EmailNotInDomainsNotInEmails (0.00s)
    --- PASS: TestValidatorSubDomains/EmailInLastEmailList (0.00s)
    --- PASS: TestValidatorSubDomains/EmailIn1stSubdomain (0.00s)
    --- PASS: TestValidatorSubDomains/EmailIn2ndSubdomain (0.00s)
    --- PASS: TestValidatorSubDomains/EmailNotInAnySubdomain (0.00s)
    --- PASS: TestValidatorSubDomains/EmailInLastSubdomain (0.00s)
PASS
```

File: validator_test.go
Comment (morarucostel) [reply validator_test.go:376]: > This makes me feel like we need to update the documentation of this, I had expected that if I wrote `.example.com` that emails to that domain and its subdomains would have been allowed, this is how our whitelist domain works IIRC 🤔

To me if I write as allowed domain `.example.com` I expect all sub-domains to be allowed and not the main domain. If I write `example.com` I expect only the domain to be allowed (considering that we can't use wildcards). The way it is now implemented it allows the end-user to say allow any sub-domain but not the main domain (if you want to include the main domain the user can add it in the list). If we go with the behaviour similar to the whitelist I think we loose this use-case.

> Secondly, do we have a test case where we do `.example.com` as an allowed domain and the email itself is `something@fooexample.com`, just to make sure that we are including the `.` in the suffix check and don't break that in the future by trying to be clever about something

Just add a new test case for the suggested scenario.

**Comment by NickMeves** [comment]
@morarucostel I have this on my radar. We'll aim to review & merge after we wrap up the v7 release we are prepping.

**Comment by morarucostel** [comment]
@NickMeves 
bringing this back to you attention

**Comment by NickMeves** [comment]
I'll need a quick reminder which PR this was, there might be a related PR I'm thinking of that couldn't merge due to security incompatibilities until a few providers merged.

This is adding leading-dot `.domain.com` syntax to `email-domains` only right? Title is making me mix this up with a reordering PR of `*` only `email-domain` priority.

Since this was first opened, there was a CVE published for leading `.` domain logic, you might want to confirm this doesn't have the issue when you rebase (or align with the new implementation in other areas if you are sharing helper methods): https://github.com/oauth2-proxy/oauth2-proxy/security/advisories/GHSA-4mf2-f3wh-gvf2

**Pull request closed without merging**
